backend/tests/test_execution_runtime_controls.py

```
@@ -0,0 +1,202 @@
+def test_execute_candidate_sql_blocks_pre_cancelled_execution_fail_closed() -> None:
+    from app.features.execution import (
+        ExecutionRuntimeCancelledError,
+        execute_candidate_sql,
+    )
+
+    def fake_query_runner(*, database_url: str, canonical_sql: str) -> list[dict[str, Any]]:
+        raise AssertionError("query runner should not run after cancellation")
+
+    with pytest.raises(
+        ExecutionRuntimeCancelledError,
+        match="Execution canceled before the backend-owned query runner started.",
```

Ruff (0.15.10)

[warning] Pattern passed to `match=` contains metacharacters but is neither escaped nor raw

(RUF043)